pydocstringformatter/_formatting/formatters_numpydoc.py

```diff
@@ -0,0 +1,85 @@
+from __future__ import annotations
+
+from pydocstringformatter._formatting.base import NumpydocSectionFormatter
+
+
+class NumpydocSectionOrderingFormatter(NumpydocSectionFormatter):
+    """Change section order to match numpydoc guidelines."""
+
+    name = "numpydoc-section-order"
+
+    numpydoc_section_order = (
+        # I'm folding summary, deprecation warning, and extended
+        # summary into "before first heading"
+        "Parameters",
+        "Attributes",
+        "Methods",
+        "Returns",
+        "Yields",
+        "Receives",
+        "Other Parameters",
+        "Raises",
+        "Warns",
+        "Warnings",
+        "See Also",
+        "Notes",
+        "References",
+        "Examples",
+    )
+
+    def treat_sections(self, sections: list[list[str]]) -> list[list[str]]:
+        """Sort the numpydoc sections into the numpydoc order."""
+        # section_names = [sec_lines[0].strip() for sec_lines in sections]
+        # assert all(name in self.numpydoc_section_order for name in section_names[1:])
+        new_sections = sorted(
+            sections[1:],
+            key=lambda sec_lines: self.numpydoc_section_order.index(
+                sec_lines[0].strip()
+            ),
+        )
+        new_sections.insert(0, sections[0])
+        return new_sections
+
+
+class NumpydocParameterNameColonTypeFormatter(NumpydocSectionFormatter):
+    """Ensure proper spacing around the colon separating names from types."""
+
+    name = "numpydoc-name-type-spacing"
+
+    numpydoc_sections_with_parameters = (
+        "Parameters",
+        "Attributes",
+        "Returns",
+        "Yields",
+        "Receives",
+        "Other Parameters",
+        "See Also",
+    )
+
+    def treat_sections(self, sections: list[list[str]]) -> list[list[str]]:
+        """Ensure proper spacing around the colon separating names from types."""
+        new_sections = []
+        for section_lines in sections:
+            section_name = section_lines[0].lstrip()
```

Comment:
This is something we tend to do a lot? Should `sections` be an `OrderedDict`? With the name as key already?

Reply:
That might work.  Alternately, look through the existing base classes to see if there's one to strip off the (common) indentation before processing and put it back after (that would also sidestep the `last_line` complexities above).

Reply:
I don't think there is currently.

Reply:
Should I make a separate one or just keep that as part of the NumpydocSectionsFormatter or whatever I called it?  Or just make `sections` an `OrderedDict` (there's no `.sortkeys(key)` method on `OrderedDict`, which means I'll need a couple of passes through the sections.

Reply:
I was thinking of letting `NumpydocSectionFormatter` pass an `OrderedDict` to `treat_sections` in its `treat_string`. `NumpydocSectionOrderingFormatter` can then return an `OrderedDict` of a different order.

Reply:
Although, since `Python` `3.6` dicts are also order dependant. So we could just rely on that fact? I'm not sure, `OrderedDict` is probably better.

Reply:
I think it's [3.7](https://docs.python.org/3/library/collections.html#collections.OrderedDict) that the `dict` ordering became a Python thing rather than a CPython and PyPy implementation detail.  `OrderedDict` makes that intention explicit, which will probably help with maintenance.